### ISSUE TITLE
Refactor/Rename's regex replace can break the code when target identifier's type contains the full target identifier

### ISSUE THREAD

**Issue by retailcoder** [retailcoder/Rubberduck#381]
This code currently determines how a code line gets rewritten when performing a rename refactoring:

        private string GetReplacementLine(string content, string target, string newName)
        {
            return Regex.Replace(content, "\\b" + target + "\\b", newName);
        }

This works very well in most cases, but breaks (doesn't work as intended) in a case like this:

    Public Sub DoSomething(ByRef connection As ADODB.connection)
        '...
    End Sub

Renaming the `connection` parameter to `foo` will correctly update all references to it in the `DoSomething` scope... but the type of the parameter will be broken:

    Public Sub DoSomething(ByRef foo As ADODB.foo)
        '...
    End Sub

Any ideas?

**Comment by ckuhn203** [comment]
Match whitespace instead. I haven't actually run this code, but it should be close.

    private string GetReplacementLine(string content, string target, string newName)
    {
        return Regex.Replace(content, "(\\s)(" + target + ")(\\s)", "$1" + newName + "$3");
    }

See [this refiddle](http://refiddle.com/28ws) for the regex pattern and [the substitutions section of this msdn doc](https://msdn.microsoft.com/en-us/library/ewy2t5e0(v=vs.110).aspx) for some clarification on the replacement string.

Basically, the .Net flavor or regex defines a period as a word break, so we need to match on whitespace instead. However, to deal with tabs/newlines, we need to wrap the replacement value in whatever whitespace we originally found around it. This is accomplished in the code above by using numbered groups. Named groups might offer a bit more clarity to the code.

**Comment by ckuhn203** [comment]
As a side note, I would probably use the string literal token and define some variables at this point.

    private string GetReplacementLine(string content, string target, string newName)
    {
        var find = @"(\s)(" + target + @")(\s)";
        var replacement = "$1" + newName + "$3";

        return Regex.Replace(content, find, replacement);
    }

**Comment by ckuhn203** [comment]
Nevermind me. I'm an idiot. This would work for this one instance, but not generally, because we could be renaming something like a public field, which will be delimited by a period in the client code.

The only way I can think to reliably do this is to find the exact location of the target using the symbols table you've been working on. Is that even remotely possible?

**Comment by retailcoder** [comment]
Yes, every `IdentifierReference` knows exactly where it is in the module. Problem is with the replacement operation itself: 

    Foo = Foo + 1

The position of the 2nd foo would be off after the first replacement.

We need to use a `TokenStreamRewriter` and regroup identifier references by code line (to avoid invalidating the symbol table between each replacement), and replace "the 2nd and 5th tokens" with the new name, *and then* push the rewritten code line to the VBE.

**Comment by retailcoder** [comment]
Commit 43a47d06e46924460d28220ca2b461e5c1b4f441 fixes this.
